src/coreComponents/physicsSolvers/fluidFlow/wells/SinglePhaseWellKernels.cpp

```diff
@@ -0,0 +1,669 @@
+  }
+  else // rate control
+  {
+    // the control is viable if the reference pressure is below/above the max/min pressure
+    if( wellType == WellControls::Type::PRODUCER )
+    {
+      // targetBHP specifies a min pressure here
+      controlIsViable = ( currentBHP >= targetBHP );
+    }
+    else
+    {
+      // targetBHP specifies a max pressure here
+      controlIsViable = ( currentBHP <= targetBHP );
+    }
+  }
+
+  if( controlIsViable )
+  {
+    newControl = currentControl;
+  }
+  else
+  {
+    // Note: if BHP control is not viable, we switch to TOTALVOLRATE
+    //       if TOTALVOLRATE are not viable, we switch to BHP
+    newControl = ( currentControl == WellControls::Control::BHP )
+               ? WellControls::Control::TOTALVOLRATE
+               : WellControls::Control::BHP;
+  }
+}
+
+GEOSX_HOST_DEVICE
+void
+ControlEquationHelper::
+  compute( globalIndex const rankOffset,
+           WellControls::Control const currentControl,
+           real64 const & targetBHP,
+           real64 const & targetRate,
+           real64 const & currentBHP,
+           real64 const & dCurrentBHP_dPres,
+           real64 const & currentVolRate,
+           real64 const & dCurrentVolRate_dPres,
+           real64 const & dCurrentVolRate_dRate,
+           globalIndex const dofNumber,
+           CRSMatrixView< real64, globalIndex const > const & localMatrix,
+           arrayView1d< real64 > const & localRhs )
+{
+  localIndex const eqnRowIndex = dofNumber + ROFFSET::CONTROL - rankOffset;
+  globalIndex const presDofColIndex = dofNumber + COFFSET::DPRES;
+  globalIndex const rateDofColIndex = dofNumber + COFFSET::DRATE;
+
+  real64 controlEqn = 0;
+  real64 dControlEqn_dRate = 0;
+  real64 dControlEqn_dPres = 0;
+
+  // Note: We assume in the computation of currentBHP that the reference elevation
+  //       is in the top well element. This is enforced by a check in the solver.
+  //       If we wanted to allow the reference elevation to be outside the top
+  //       well element, it would make more sense to check the BHP constraint in
+  //       the well element that contains the reference elevation.
+
+  // BHP control
+  if( currentControl == WellControls::Control::BHP )
+  {
+    // control equation is a difference between current BHP and target BHP
+    controlEqn = currentBHP - targetBHP;
+    dControlEqn_dPres = dCurrentBHP_dPres;
+  }
+  // Total volumetric rate control
+  else if( currentControl == WellControls::Control::TOTALVOLRATE )
+  {
+    // control equation is the difference between volumetric current rate and target rate
+    controlEqn = currentVolRate - targetRate;
+    dControlEqn_dRate = dCurrentVolRate_dRate;
+    dControlEqn_dPres = dCurrentVolRate_dPres;
+  }
+  else
+  {
+    GEOSX_ERROR_IF( ( currentControl != WellControls::Control::BHP )
+                    && ( currentControl != WellControls::Control::TOTALVOLRATE ),
```

Comment:
These conditions seem unnecessary in the final `else` branch

Reply:
I removed the condition here.